ray-operator/controllers/ray/batchscheduler/koordinator/koordinator_scheduler_test.go

```diff
@@ -0,0 +1,103 @@
+package koordinator
+
+import (
+	"context"
+	"testing"
+
+	rayv1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+	"github.com/stretchr/testify/assert"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func createRayClusterWithLabels(namespace, name string, labels map[string]string) *rayv1.RayCluster {
+	rayCluster := &rayv1.RayCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+			Labels:    labels,
+		},
+	}
+
+	return rayCluster
+}
+
+func setHeadPodNamespace(app *rayv1.RayCluster, namespace string) {
+	app.Spec.HeadGroupSpec.Template.Namespace = namespace
+}
+
+func addWorkerPodSpec(app *rayv1.RayCluster, workerGroupName string,
+	namespace string, replicas, minReplicas int32) {
+	workerGroupSpec := rayv1.WorkerGroupSpec{
+		GroupName:   workerGroupName,
+		Replicas:    &replicas,
+		MinReplicas: &minReplicas,
+	}
+	workerGroupSpec.Template.Namespace = namespace
+
+	app.Spec.WorkerGroupSpecs = append(app.Spec.WorkerGroupSpecs, workerGroupSpec)
+}
+
+func createPod(namespace, name string) *v1.Pod {
+	return &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   namespace,
+			Labels:      make(map[string]string),
+			Annotations: make(map[string]string),
+		},
+	}
+}
+
+func TestAddMetadataToPod(t *testing.T) {
+	ks := &KoodinatorScheduler{}
+	ctx := context.Background()
+	// test the case when gang-scheduling is enabled
+	rayClusterWithGangScheduling := createRayClusterWithLabels(
+		"ray-namespace",
+		"koord",
+		map[string]string{
+			utils.RayClusterGangSchedulingEnabled: "true",
+		},
+	)
+
+	setHeadPodNamespace(rayClusterWithGangScheduling, "ns0")
```

Comment:
Why is this call needed? The namespace should be inherited from the RayCluster namespace right?

Reply:
> Why is this call needed? The namespace should be inherited from the RayCluster namespace right?

What you said is absolutely correct. However, to make this module's code more general, I considered that there might be scenarios in the future where the namespace for head pods or worker pods needs to be specifically designated. Therefore, I implemented compatibility here: if the namespace for the head pod or worker pod is empty, it will inherit the namespace of the rayCluster; otherwise, it will be processed according to the specified namespace.

Of course, if you think we don’t need to consider such special cases for now, I completely agree as well.